### PR TITLE
More user-friendly API.

### PR DESCRIPTION
**Pull request by clarcharr**
I was trying to use this crate and stumbled upon the API. I figure that it's easier to just offer a redesign and ask for critique.

Basically, the changes are:

* made `cycles` not public because I don't think it was supposed to be
* renamed `Config` to `Builder` so that it makes a bit more sense
* removed `Ratelimit::new` because it was confusing; it makes more sense to say `Ratelimit::default`
* removed `Ratelimit::configure` and replaced it with `Builder::new`
* renamed `block` to `wait_for` and added `wait` which is just `wait_for(1)`
* wrapped `clone_sender` into an opaque `Handle` type with methods `wait` and `try_go` instead of `send` and `try_send`
* renamed `Ratelimit` to `Limiter`
* made `run` private and encapsulated it in a `SyncLimiter`, which spawns a `run` thread and offers the creation of `Handle`

It may be easier to understand the changes by just running `cargo doc` and viewing the changes directly. Let me know what you think of the changes!

A few other potential things that could be done to the docs:

1. Clarify exactly what quantum/capacity/interval actually mean, explaining the algorithm more in detail. I understand this but can't really fully word it right now.
2. Specify a few use case examples for configurations that make use of the above.

PS: thanks for the great crate!

### CODE REVIEW

**Comment by brayniac** [comment]
Hi @clarcharr - thanks for your PR! I definitely want to make sure the library's API makes it easy to use - and admit there's definitely room for improvements. Let me take some time to review the changes in more detail so I can offer some meaningful review.

File: src/lib.rs [hunk 1]
```diff
@@ -422,7 +499,7 @@ mod tests {
     #[cfg(feature = "unstable")]
     #[bench]
     fn block_1_million_per_second_1000ns(b: &mut test::Bencher) {
-        let mut r = Ratelimit::configure()
+        let mut r = Limiter::configure()
```
Comment (brayniac) [inline src/lib.rs]: Need a Builder here

File: src/lib.rs [hunk 2]
```diff
@@ -432,7 +509,7 @@ mod tests {
     #[cfg(feature = "unstable")]
     #[bench]
     fn block_2_million_per_second_500ns(b: &mut test::Bencher) {
-        let mut r = Ratelimit::configure()
+        let mut r = Limiter::configure()
```
Comment (brayniac) [inline src/lib.rs]: Need a Builder

File: src/lib.rs [hunk 3]
```diff
@@ -442,7 +519,7 @@ mod tests {
     #[cfg(feature = "unstable")]
     #[bench]
     fn block_10_million_per_second_100ns(b: &mut test::Bencher) {
-        let mut r = Ratelimit::configure()
+        let mut r = Limiter::configure()
```
Comment (brayniac) [inline src/lib.rs]: Need a Builder

File: src/lib.rs [hunk 4]
```diff
@@ -452,7 +529,7 @@ mod tests {
     #[cfg(feature = "unstable")]
     #[bench]
     fn block_20_million_per_second_50ns(b: &mut test::Bencher) {
-        let mut r = Ratelimit::configure()
+        let mut r = Limiter::configure()
```
Comment (brayniac) [inline src/lib.rs]: Need a Builder

File: src/lib.rs [hunk 5]
```diff
@@ -175,40 +232,69 @@ impl Default for Config {
...
+    done: Arc<atomic::AtomicBool>,
+    handle: Handle,
+}
+impl SyncLimiter {
+    /// Gets a handle to this rate limiter.
     ///
     /// # Example
     /// ```
-    /// # use ratelimit::Ratelimit;
+    /// let multi = SyncLimiter::default();
```
Comment (brayniac) [inline src/lib.rs]: Scope issue here in the tests?

File: src/lib.rs [hunk 6]
```diff
@@ -175,40 +232,69 @@ impl Default for Config {
...
     ///
-    /// let mut r = Ratelimit::new();
-    pub fn new() -> Ratelimit {
-        Default::default()
+    /// for i in 0..2 {
+    ///     let s = multi.make_handle();
+    ///     thread::spawn(move || {
```
Comment (brayniac) [inline src/lib.rs]: Scope issue here too?

File: src/lib.rs [hunk 7]
```diff
@@ -73,100 +66,164 @@ extern crate shuteye;
...
+        self.inner.send(()).unwrap_or_else(|_| unreachable!());
     }
 
-    /// sets the number of tokens to add per interval
+    /// Checks if we would be able to get a token, and returns whether we got it.
     ///
     /// # Example
     /// ```
-    /// # use ratelimit::Ratelimit;
+    /// let multi = SyncLimiter::default();
```
Comment (brayniac) [inline src/lib.rs]: SyncLimiter not in-scope?

File: src/lib.rs [hunk 8]
```diff
@@ -73,100 +66,164 @@ extern crate shuteye;
...
+    inner: mpsc::SyncSender<()>,
+}
+unsafe impl Sync for Handle {}
+impl Handle {
+    /// Waits for a single token.
     ///
     /// # Example
     /// ```
-    /// # use ratelimit::Ratelimit;
+    /// let multi = SyncLimiter::default();
```
Comment (brayniac) [inline src/lib.rs]: Scope issue?

File: src/lib.rs [hunk 9]
```diff
@@ -73,100 +66,164 @@ extern crate shuteye;
...
     ///
-    /// let mut r = Ratelimit::configure().build();
-    pub fn build(self) -> Ratelimit {
-        Ratelimit::configured(self)
+    /// for i in 0..2 {
+    ///     let s = multi.make_handle();
+    ///     thread::spawn(move || {
```
Comment (brayniac) [inline src/lib.rs]: Scope issue?

**Comment by brayniac** [comment]
@clarcharr I like the changes. Seems like it's really going to make it much cleaner to use the library.

I found https://github.com/blankenshipz/oanda-rs/blob/master/src/client.rs which makes use of the multi-threaded version. Looks like the changes make sense within the context of that use-case.

Once the tests are passing, I'm happy to merge and cut 0.5.0

**Comment by clarcharr** [comment]
So I thought that I ran `cargo test` and it passed but I probably missed a lot. I'll fix these later today.

**Comment by clarcharr** [comment]
One other important change I made I figure I should point out: for a while I'd been adding `deny(warnings)` to my crates too, but it makes a lot more sense to do `cfg_attr(test, deny(warnings))` to future-proof the crate from new lints being added.

So I updated that for this crate while I was at it.

**Comment by clarcharr** [comment]
(This should all work, and everything is ready to merge. I'd appreciate if you took another look at the rustdoc just to make sure everything makes sense before you do though.)

**Comment by brayniac** [comment]
@clarcharr - I suspect you need to lower the capacity - as that sets the capacity of the mpsc::SyncSender. Since we're using the queue backwards, I suspect it's able to push the token onto the queue, and therefor doesn't block on call to wait(). Reducing the capacity to 1 might make sense for these tests?

**Comment by clarcharr** [comment]
That makes sense for the tests but what I'm more concerned about is the fact that `Handle::wait` and `Limiter::wait` have different functionality and I'm not quite sure how to reflect that in the names/docs.

**Comment by brayniac** [comment]
Yeah, this actually makes describing the functionality challenging. I wonder if we should refactor to use a MPMC queue in the correct direction. crates.io/crates/mpmc is pretty fast. 

Thanks for uncovering this issue though. Correctness is important, and it seems off in the current state. =)

**Comment by clarcharr** [comment]
Seems like a good idea! I don't mind merging what's there now (as it replicates all the existing functionality under a different API) and then modifying the underlying implementation later. The nice thing about using opaque types is that the underlying implementation can be changed without actually changing the API.

I mean, the only thing that *feels* wrong to me is that `Handle`s can be used after a `RateLimiter` is dropped. I did document this behaviour for now, and this could be solved by mucking around with `PhantomData` to add extra lifetimes, but I'm not necessarily sure if that's a reasonable path to go down.

**Comment by brayniac** [comment]
Totally. That was broken before. I think going back to the "normal" direction for the SyncLimiter is the most reasonable course. That should address the behavior if the RateLimiter is dropped too? (mpmc queue would back-up with tokens, and wait() would block indefinitely?)

**Comment by clarcharr** [comment]
@brayniac that one is actually solved by the fact that creating a `SyncLimiter` consumes the original `Limiter`, so, the original limiter will *only* be dropped if the `SyncLimiter` is dropped.

This is because when the `SyncLimiter` is dropped, an `Arc<Atomic>` bool is switched to false. This will get caught by the `while` loop in the running thread, signalling the thread to exit, dropping the original limiter.

**Comment by brayniac** [comment]
Planning to merge this evening. I want to take one more look at the rustdoc first.

**Comment by brayniac** [comment]
Merged. Thanks @clarcharr for your submission!

**Pull request merged**